rules/aip0136/response_message_name.go

```diff
@@ -0,0 +1,69 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package aip0136
+
+import (
+	"fmt"
+
+	"github.com/googleapis/api-linter/lint"
+	"github.com/googleapis/api-linter/locations"
+	"github.com/googleapis/api-linter/rules/internal/utils"
+	"github.com/jhump/protoreflect/desc"
+)
+
+// Custom methods should return a response message matching the RPC name,
+// with a Response suffix, or the resource being operated on.
+var responseMessageName = &lint.MethodRule{
+	Name:   lint.NewRuleName(136, "response-message-name"),
+	OnlyIf: isCustomMethod,
+	LintMethod: func(m *desc.MethodDescriptor) []lint.Problem {
+		// A response is considered valid if
+		// - The response name matches the RPC name with a `Response` suffix
+		// - The response is the resource being operated on
+		// To identify the resource being operated on, we inspect the Resource
+		// Reference of the input type's `name` field. This guidance is documented
+		// in https://google.aip.dev/136#resource-based-custom-methods
+
+		response := utils.GetResponseType(m)
+		requestType := utils.GetResourceReference(m.GetInputType().FindFieldByName("name")).GetType()
+		resourceOperatedOn := utils.FindResourceMessage(requestType, m.GetFile())
```

Comment:
We don't technically need to do a message type look up - we can simply match the resource_reference type to the response's `google.api.resource` type. So a slightly simpler thing to do would be:

```
response := utils.GetResponseType(m)
requestResourceType := utils.GetResourceReference(m.GetInputType().FindFieldByName("name")).GetType()

if !utils.IsResource(response) || requestResourceType == "" {
  return suffixFindings // from the "lint method has matching response" call moved earlier, see other comment
}

responseResourceType := utils.GetResource(response).GetType()
if responseResourceType == requestResourceType {
  // all good!
  return nil
}

// return findings defined below for the either-or guidance violation
```

This also reads straight down a bit better, exiting early whenever possible.

Note: in testing you will need to add a resource annotation on the response message

Reply:
We need the message however for a suggestion.

According to https://google.aip.dev/client-libraries/4231#resource-messages

> The universal resource type usually (but not always) ends with a name that matches the name of the message.

So I figured, because I needed to fetch the message type from the file for the suggestion, it could serve both purposes without having to do the type match.  I can still do the type match as well for defensive purposes if you think it'll help?

Reply:
See later comment re:the `Suggestion` :)

Reply:
Yeah, replied before I got through all the comments haha :)

Addressed this here: https://github.com/googleapis/api-linter/pull/1387/commits/b007953dcd47b31c661c4f18e1694236715c3e4b

However, I didn't want to return the `suffixFindings` because I wanted to make sure the error message was the same through all of the situations so that consumers understood it could be either or, and in no situation should we suggest something... yet at least.

Reply:
> However, I didn't want to return the suffixFindings because I wanted to make sure the error message was the same through all of the situations so that consumers understood it could be either or, and in no situation should we suggest something... yet at least.

Fair enough, I agree, that's fine by me!

Reply:
Cleaning up one more typo in my next commit, then should be GTG for a final pass

Reply:
Feel free to hit the "re-request review" button whenever: 
![image](https://github.com/googleapis/api-linter/assets/6644735/6a0770d9-d63e-41e4-ade2-7df72ea0ce5d)